docarray/document/mixins/multimodal.py

```diff
@@ -0,0 +1,134 @@
+                                    f'Unsupported type annotation inside Iterable: {sub_type}'
+                                )
+                            chunk.chunks.append(doc)
+                        multi_modal_schema[key] = {
+                            'attribute_type': attribute_type,
+                            'type': f'{field.type._name}[{sub_type.__name__}]',
+                            'position': len(root.chunks),
+                        }
+                        root.chunks.append(chunk)
+                else:
+                    raise ValueError(f'Unsupported type annotation {field.type._name}')
+            else:
+                doc, attribute_type = cls._from_obj(attribute, field.type)
+                multi_modal_schema[key] = {
+                    'attribute_type': attribute_type,
+                    'type': field.type.__name__,
+                    'position': len(root.chunks),
+                }
+                root.chunks.append(doc)
+
+        # TODO: may have to modify this?
+        root.tags = tags
+        root._metadata['multi_modal_schema'] = multi_modal_schema
+
+        return root
+
+    def get_multi_modal_attribute(self, attribute: str) -> 'DocumentArray':
+        from docarray import DocumentArray
+
+        if 'multi_modal_schema' not in self._metadata:
+            raise ValueError(
+                'the Document does not correspond to a Multi Modal Document'
+            )
+
+        if attribute not in self._metadata['multi_modal_schema']:
+            raise ValueError(
+                f'the Document schema does not contain attribute {attribute}'
+            )
+
+        attribute_type = self._metadata['multi_modal_schema'][attribute][
+            'attribute_type'
+        ]
+        position = self._metadata['multi_modal_schema'][attribute].get('position')
+
+        if attribute_type in [AttributeType.DOCUMENT, AttributeType.NESTED]:
+            return DocumentArray([self.chunks[position]])
+        elif attribute_type in [
+            AttributeType.ITERABLE_DOCUMENT,
+            AttributeType.ITERABLE_NESTED,
+        ]:
+            return self.chunks[position].chunks
+        else:
+            raise ValueError(
+                f'Invalid attribute {attribute}: must a Document attribute or nested dataclass'
+            )
+
+    @classmethod
+    def _from_obj(cls, obj, obj_type) -> typing.Tuple['Document', AttributeType]:
+        from docarray import Document
+
+        attribute_type = AttributeType.DOCUMENT
+
+        if is_dataclass(obj_type):
+            doc = cls.from_dataclass(obj)
+            attribute_type = AttributeType.NESTED
+        elif obj_type == ImageDocument:
+            doc = Document(tensor=obj, modality='image')
+        elif obj_type == BlobDocument:
+            doc = Document(blob=obj, modality='blob')
+        elif obj_type == TextDocument:
+            doc = Document(text=obj, modality='text')
+        else:
```

Comment:
add `elif obj_type == Document`?

Reply:
the typing module doesn't offer a `Document` type annotation.
If we're supposed to offer such an annotation, how do you suggest it should work ?

Reply:
I'm also not quite clear about it. Does it make sense to offer an annotation of `Document`?

Reply:
I don't think so, I think the modality of the type should be clear.
We can also add a type annotation `Field` to allow users to provide their custom types but I'm not sure about the priority for it